muchsync.1.md
% muchsync(1)
% David Mazieres
% 

# NAME

muchsync - synchronize maildirs and notmuch databases

# SYNOPSIS

muchsync _options_ \
muchsync _options_ _server-name_ _server-options_ \
muchsync _options_ --init _maildir_ _server-name_ _server-options_

# DESCRIPTION

muchsync synchronizes the contents of maildirs and notmuch tags across
machines.  Any given execution runs pairwise between two replicas, but
the system scales to an arbitrary number of replicas synchronizing in
arbitrary pairs.  For efficiency, version vectors and logical
timestamps are used to limit synchronization to items a peer may not
yet know about.

To use muchsync, both muchsync and notmuch should be installed
someplace in your PATH on two machines, and you must be able to access
the remote machine via ssh.

In its simplest usage, you have a single notmuch database on some
server `SERVER` and wish to start replicating that database on a
client, where the client currently does not have any mailboxes.  You
can initialize a new replica in `$HOME/inbox` by running the following
command:

    muchsync --init $HOME/inbox SERVER

This command may take some time, as it transfers the entire contents
of your maildir from the server to the client and creates a new
notmuch index on the client.  Depending on your setup, you may be
either bandwidth limited or CPU limited.  (Sadly, the notmuch library
on which muchsync is built is non-reentrant and forces all indexing to
happen on a single core at a rate of about 10,000 messages per
minute.)

From then on, to synchronize the client with the server, just run:

    muchsync SERVER

Since muchsync replicates the tags in the notmuch database itself, you
should consider disabling maildir flag synchronization by executing:

    notmuch config set maildir.synchronize_flags=false

The reason is that the synchronize\_flags feature only works on a
small subset of pre-defined flags and so is not all that useful.
Moreover, it marks flags by renaming files, which is not particularly
efficient.  muchsync was largely motivated by the need for better flag
synchronization.  If you are satisfied with the synchronize\_flags
feature, you might consider a tool such as offlineimap as an
alternative to muchsync.


## Synchronization algorithm

muchsync separately synchronizes two classes of information:  the
message-to-directory mapping (henceforth link counts) and the
message-id-to-tag mapping (henceforth tags).  Using logical
timestamps, it can detect update conflicts for each type of
information.  We describe link count and tag synchronization in turn.

Link count synchronization consists of ensuring that any given message
(identified by its collision-resistant content hash) appears the same
number of times in the same subdirectories on each replica.  Generally
a message will appear only once in a single subdirectory.  However, if
the message is moved or deleted on one replica, this will propagate to
other replicas.

If two replicas move or copy the same file between synchronization
events (or one moves the file and the other deletes it), this
constitutes an update conflict.  Update conflicts are resolved by
storing in each subdirectory a number of copies equal to the maximum
of the number of copies in that subdirectory on the two replicas.
This is conservative, in the sense that a file will never be deleted
after a conflict, though you may get extra copies of files.  (muchsync
uses hard links, so at least these copies will not use too much disk
space.)

For example, if one replica moves a message to subdirectory .box1/cur
and another moves the same message to subdirectory .box2/cur, the
conflict will be resolved by placing two links to the message on each
replica, one in .box1/cur and one in .box2/cur.  To respect the
structure of maildirs, subdirectories ending `new` and `cur` are
special-cased; conflicts between sibling `new` and `cur`
subdirectories are resolved in favor of `cur` without creating
additional copies of messages.

Message tags are synchronized based on notmuch's message-ID (usually
the Message-ID header of a message), rather than message contents.  On
conflict, tags are combined as follows.  Any tag in the notmuch
configuration parameter `new.tags` is removed from the message unless
it appears on both replicas.  Any other tag is added if it appears on
any replica.  In other words, tags in `new.tags` are logically anded,
while all other flags are logically ored.  (This approach will give
the most predictable results if `new.tags` has the same value in all
your replicas.  The `--init` option ensures this initially, but
subsequent changes to `new.tags` must be manually propagated.)

## File deletion

Because publishing software that actually deletes people's email is a
scary prospect, muchsync for the moment never actually deletes mail
files.  Though this may change in the future, for the moment muchsync
moves any deleted messages to the directory `.notmuch/muchsync/trash`
under your mail directory (naming deleted messages by their content
hash).  If you really want to delete mail to reclaim disk space or for
privacy reasons, you will need to run the following on each replica:

    cd "$(notmuch config get database.path)"
    rm -rf .notmuch/muchsync/trash

# OPTIONS

\-C _file_, \--config _file_
:   Specify the path of the notmuch configuration file to use.  If
    none is specified, the default is to use the contents of the
    environment variable \$NOTMUCH_CONFIG, or if that variable is
    unset, the value \$HOME/.notmuch-config.  (These are the same
    defaults as the notmuch command itself.)

\-F
:   Check for modified files.  Without this option, muchsync assumes
    that files in a maildir are never edited.  -F disables certain
    optimizations so as to make muchsync at least check the timestamp
    on every file, which will detect modified files at the cost of a
    longer startup time.

\-r /path/to/muchsync
:   Specifies the path to muchsync on the server.  Ordinarily, muchsync
    should be in the default PATH on the server so this option is not
    required.  However, this option is useful if you have to install
    muchsync in a non-standard place or wish to test development
    versions of the code.

\-s ssh-cmd
:   Specifies a command line to pass to /bin/sh to execute a command on
    another machine.  The default value is "ssh -CTaxq".  Note that
    because this string is passed to the shell, special characters
    including spaces may need to be escaped.

\-v
:   The -v option increases verbosity.  The more times it is specified,
    the more verbose muchsync will become.

\--help
:   Print a brief summary of muchsync's command-line options.

\--init _maildir_
:   This option clones an existing mailbox on a remote server into
    _maildir_ on the local machine.  Neither _maildir_ nor your
    notmuch configuration file (see ```--config``` above) should exist
    when you run this command, as both will be created.  The
    configuration file is copied from the server (adjusted to reflect
    the local maildir), while _maildir_ is created as a replica of the
    maildir you have on the server.

\--nonew
:   Ordinarily, muchsync begins by running "notmuch new".  This option
    says not to run "notmuch new" before starting the muchsync
    operation.  It can be passed as either a client or a server
    option.  For example:  The command "```muchsync myserver
    --nonew```" will run "```notmuch new```" locally but not on
    myserver.

\--noup, \--noupload
:   Transfer files from the server to the client, but not vice versa.

\--upbg
:   Transfer files from the server to the client in the foreground.
    Then fork into the background to upload any new files from the
    client to the server.  This option is useful when checking new
    mail, if you want to begin reading your mail as soon as it has
    been downloaded while the upload continues.

\--self
:   Print the 64-bit replica ID of the local maildir replica and exit.
    Potentially useful in higher-level scripts, such as the emacs
    notmuch-poll-script variable for identifying on which replica one
    is running, particularly if network file systems allow a replica
    to be accessed from multiple machines.

\--version
:   Report on the muchsync version number

# EXAMPLES

To initialize a the muchsync database, you can run:

    muchsync -vv

This first executes "`notmuch new`", then builds the initial muchsync
database from the contents of your maildir (the directory specified as
`database.path` in your notmuch configuration file).  This command may
take several minutes the first time it is run, as it must compute a
content hash of every message in the database.  Note that you do not
need to run this command, as muchsync will initialize the database the
first time a client tries to synchronize anyway.

    muchsync --init ~/maildir myserver

First run "notmuch new" on myserver, then create a directory
`~/maildir` containing a replica of your mailbox on myserver.  Note
that neither your configuration file (by default `~/.notmuch-config`)
nor `~/maildir` should exist before running this command, as both will
be created.

To create a `notmuch-poll` script that fetches mail from a remote
server `myserver`, but on that server just runs `notmuch new`, do the
following:  First, run `muchsync --self` on the server to get the
replica ID.  Then take the ID returned (e.g., `1968464194667562615`)
and embed it in a shell script as follows:

    #!/bin/sh
    self=$($HOME/muchsync --self) || exit 1
    if [ "$self" = 1968464194667562615 ]; then
        exec notmuch new
    else
        exec $HOME/muchsync -r ./muchsync --upbg myserver
    fi

The path of such a script is a good candidate for the emacs
`notmuch-poll-script` variable.

Alternatively, to have the command ``notmuch new`` on a client
automatically fetch new mail from server `myserver`, you can place the
following in the file ``.notmuch/hooks/post-new`` under your mail
directory:

    #!/bin/sh
    notmuch --nonew --upbg myserver

# FILES

The default notmuch configuration file is `$HOME/.notmuch-config`.

muchsync keeps all of its state in a subdirectory of your top maildir
called ```.notmuch/muchsync```.

# SEE ALSO

notmuch(1).

# BUGS

muchsync expects initially to create replicas from scratch.  If you
have created a replica using another tool such as offlineimap and you
try to use muchsync to synchronize them, muchsync will assume every
file has an update conflict.  This is okay if the two replicas are
identical; if they are not, it will result in artifacts such as files
deleted in only one replica reappearing.  Ideally notmuch needs an
option like `--clobber` that makes a local replica identical to the
remote one without touching the remote one, so that an old version of
a mail directory can be used as a disposable cache to bootstrap
initialization.

muchsync never deletes directories.  If you want to remove a
subdirectory completely, you must manually execute rmdir on all
replicas.  Even if you manually delete a subdirectory, it will live on
in the notmuch database.

To synchronize deletions and re-creations properly, muchsync never
deletes content hashes and their message IDs from its database, even
after the last copy of a message has disappeared.  Such stale hashes
should not consume an inordinate amount of disk space, but could
conceivably pose a privacy risk if users believe deleting a message
removes all traces of it.

Message tags are synchronized based on notmuch's message-ID (usually
the Message-ID header of a message), rather than based on message
contents.  This is slightly strange because very different messages
can have the same Message-ID header, meaning the user will likely only
read one of many messages bearing the same Message-ID header.  It is
conceivable that an attacker could suppress a message from a mailing
list by sending another message with the same Message-ID.  This bug is
in the design of notmuch, and hence not something that muchsync can
work around.  muchsync itself does not assume Message-ID equivalence,
relying instead on content hashes to synchronize link counts.  Hence,
any tools used to work around the problem should work on all replicas.

Because notmuch and Xapian do not keep any kind of modification time
on database entries, every invocation of muchsync requires a complete
scan of all tags in the Xapian database to detect any changed tags.
Fortunately muchsync heavily optimizes the scan so that it should take
well under a second for 100,000 mail messages.  However, this means
that interfaces such as those used by notmuch-dump are not efficient
enough (see the next paragraph).

muchsync makes certain assumptions about the structure of notmuch's
private types `notmuch_message_t` and `notmuch_directory_t`.  In
particular, it assumes that the Xapian document ID is the second field
of these data structures.  Sadly, there is no efficient and clean way
to extract this information from the notmuch library interface.
muchsync also makes other assumptions about how tokens are named in
the Xapian database.  These assumptions are necessary because the
notmuch library interface and the notmuch dump utility are too slow to
support synchronization every time you check mail.



muchsync.cc

#include <cstring>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <fcntl.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "misc.h"
#include "muchsync.h"
#include "infinibuf.h"

using namespace std;

#if 0
// This gives core dumps to make it easier to debug
struct no_such_exception_t {
  const char *what() noexcept { return "no such exception"; }
};
using whattocatch_t = no_such_exception_t;
#else
using whattocatch_t = const exception;
#endif

#define MUCHSYNC_DEFDIR "/.notmuch/muchsync"
const char muchsync_defdir[] = MUCHSYNC_DEFDIR;
const char muchsync_dbpath[] = MUCHSYNC_DEFDIR "/state.db";
const char muchsync_trashdir[] = MUCHSYNC_DEFDIR "/trash";
const char muchsync_tmpdir[] = MUCHSYNC_DEFDIR "/tmp";

constexpr char shell[] = "/bin/sh";

bool opt_fullscan;
bool opt_noscan;
bool opt_init;
bool opt_server;
bool opt_upbg;
bool opt_noup;
bool opt_nonew;
int opt_verbose;
int opt_upbg_fd = -1;
string opt_ssh = "ssh -CTaxq";
string opt_remote_muchsync_path = "muchsync";
string opt_notmuch_config;
string opt_init_dest;

static bool
muchsync_init (const string &maildir, bool create = false)
{
  string trashbase = maildir + muchsync_trashdir + "/";
  if (!access ((maildir + muchsync_tmpdir).c_str(), 0)
      && !access ((trashbase + "ff").c_str(), 0))
    return true;

  if (create && mkdir (maildir.c_str(), 0777) && errno != EEXIST) {
    perror (maildir.c_str());
    return false;
  }

  string notmuchdir = maildir + "/.notmuch";
  if (create && access (notmuchdir.c_str(), 0) && errno == ENOENT) {
    notmuch_database_t *notmuch;
    if (!notmuch_database_create (maildir.c_str(), &notmuch))
      notmuch_database_destroy (notmuch);
  }

  string msdir = maildir + muchsync_defdir;
  for (string d : {msdir, maildir + muchsync_trashdir,
	maildir + muchsync_tmpdir}) {
    if (mkdir (d.c_str(), 0777) && errno != EEXIST) {
      perror (d.c_str());
      return false;
    }
  }

  for (int i = 0; i < 0x100; i++) {
    ostringstream os;
    os << trashbase << hex << setfill('0') << setw(2) << i;
    if (mkdir (os.str().c_str(), 0777) && errno != EEXIST) {
      perror (os.str().c_str());
      return false;
    }
  }
  return true;
}

static void
tag_stderr(string tag)
{
  infinistreambuf *isb =
    new infinistreambuf(new infinibuf_mt);
  streambuf *err = cerr.rdbuf(isb);
  thread t ([=]() {
      istream in (isb);
      ostream out (err);
      string line;
      while (getline(in, line))
	out << tag << line << endl;
    });
  t.detach();
  cerr.rdbuf(isb);
}

[[noreturn]] void
usage (int code = 1)
{
  (code ? cerr : cout) << "\
usage: muchsync\n\
       muchsync server [server-options]\n\
       muchsync --init maildir server [server-options]\n\
\n\
Additional options:\n\
   -C file       Specify path to notmuch config file\n\
   -F            Disable optimizations and do full maildir scan\n\
   -v            Increase verbosity\n\
   -r path       Specify path to notmuch executable on server\n\
   -s ssh-cmd    Specify ssh command and arguments\n\
   --config file Specify path to notmuch config file (same as -C)\n\
   --nonew       Do not run notmuch new first\n\
   --noup[load]  Do not upload changes to server\n\
   --upbg        Download mail in forground, then upload in background\n\
   --self        Print local replica identifier and exit\n\
   --version     Print version number and exit\n\
   --help        Print usage\n";
  exit (code);
}

static void
print_self()
{
  unique_ptr<notmuch_db> nmp;
  try {
    nmp.reset(new notmuch_db (opt_notmuch_config));
  } catch (whattocatch_t e) { cerr << e.what() << '\n'; exit (1); }
  notmuch_db &nm = *nmp;

  string dbpath = nm.maildir + muchsync_dbpath;

  sqlite3 *db = dbopen(dbpath.c_str(), false);
  if (!db)
    exit(1);
  cleanup _c (sqlite3_close_v2, db);
  cout << getconfig<i64>(db, "self") << '\n';
}

static void
server()
{
  ifdinfinistream ibin(0);
  cleanup _fixbuf ([](streambuf *sb){ cin.rdbuf(sb); },
		   cin.rdbuf(ibin.rdbuf()));
  tag_stderr("[SERVER] ");

  unique_ptr<notmuch_db> nmp;
  try {
    nmp.reset(new notmuch_db (opt_notmuch_config));
  } catch (whattocatch_t e) { cerr << e.what() << '\n'; exit (1); }
  notmuch_db &nm = *nmp;

  string dbpath = nm.maildir + muchsync_dbpath;

  if (!opt_nonew)
    nm.run_new();
  if (!muchsync_init (nm.maildir))
    exit (1);

  sqlite3 *db = dbopen(dbpath.c_str());
  if (!db)
    exit(1);
  cleanup _c (sqlite3_close_v2, db);

  try {
    if (!opt_noscan)
      sync_local_data(db, nm.maildir);
    muchsync_server(db, nm);
  }
  catch (whattocatch_t &e) {
    cerr << e.what() << '\n';
    exit(1);
  }
}


static void
cmd_iofds (int fds[2], const string &cmd)
{
  int ifds[2], ofds[2];
  if (pipe (ifds))
    throw runtime_error (string ("pipe: ") + strerror (errno));
  if (pipe (ofds)) {
    close (ifds[0]);
    close (ifds[1]);
    throw runtime_error (string ("pipe: ") + strerror (errno));
  }

  pid_t pid = fork();
  switch (pid) {
  case -1:
    close (ifds[0]);
    close (ifds[1]);
    close (ofds[0]);
    close (ofds[1]);
    throw runtime_error (string ("fork: ") + strerror (errno));
    break;
  case 0:
    close (ifds[0]);
    close (ofds[1]);
    if (ofds[0] != 0) {
      dup2 (ofds[0], 0);
      close (ofds[0]);
    }
    if (ifds[1] != 1) {
      dup2 (ifds[1], 1);
      close (ifds[1]);
    }
    execl (shell, shell, "-c", cmd.c_str(), nullptr);
    cerr << shell << ": " << strerror (errno) << '\n';
    _exit (1);
    break;
  default:
    close (ifds[1]);
    close (ofds[0]);
    fcntl (ifds[0], F_SETFD, 1);
    fcntl (ofds[1], F_SETFD, 1);
    fds[0] = ifds[0];
    fds[1] = ofds[1];
    break;
  }
}

static void
create_config(istream &in, ostream &out, string &maildir)
{
  if (!maildir.size() || !maildir.front())
    throw runtime_error ("illegal empty maildir path\n");
  string line;
  out << "conffile\n";
  get_response(in, line);
  get_response(in, line);
  size_t len = stoul(line.substr(4));
  if (len <= 0)
    throw runtime_error ("server did not return configuration file\n");
  string conf;
  conf.resize(len);
  if (!in.read(&conf.front(), len))
    throw runtime_error ("cannot read configuration file from server\n");

  int fd = open(opt_notmuch_config.c_str(), O_CREAT|O_TRUNC|O_WRONLY|O_EXCL,
		0666);
  if (fd < 0)
    throw runtime_error (opt_notmuch_config + ": " + strerror (errno));
  write(fd, conf.c_str(), conf.size());
  close(fd);

  if (maildir[0] != '/') {
    const char *p = getenv("PWD");
    if (!p)
      throw runtime_error ("no PWD in environment\n");
    maildir = p + ("/" + maildir);
  }

  notmuch_db nm (opt_notmuch_config);
  nm.set_config ("database.path", maildir.c_str(), nullptr);
}

static void
client(int ac, char **av)
{
  unique_ptr<notmuch_db> nmp;
  struct stat sb;
  int err = stat(opt_notmuch_config.c_str(), &sb);
  if (opt_init) {
    if (!err) {
      cerr << opt_notmuch_config << " should not exist with --init option\n";
      exit (1);
    }
    else if (errno != ENOENT) {
      cerr << opt_notmuch_config << ": " << strerror(errno) << '\n';
      exit (1);
    }
  }
  else if (err) {
    cerr << opt_notmuch_config << ": " << strerror(errno) << '\n';
    exit (1);
  }
  else {
    try {
      nmp.reset(new notmuch_db (opt_notmuch_config));
    } catch (whattocatch_t e) { cerr << e.what() << '\n'; exit (1); }
  }

  if (ac == 0) {
    if (!nmp)
      usage();
    if (!muchsync_init(nmp->maildir, true))
      exit (1);
    if (!opt_nonew)
      nmp->run_new();
    string dbpath = nmp->maildir + muchsync_dbpath;
    sqlite3 *db = dbopen(dbpath.c_str());
    if (!db)
      exit (1);
    cleanup _c (sqlite3_close_v2, db);
    sync_local_data (db, nmp->maildir);
    exit(0);
  }

  ostringstream os;
  os << opt_ssh << ' ' << av[0] << ' ' << opt_remote_muchsync_path
     << " --server";
  for (int i = 1; i < ac; i++)
    os << ' ' << av[i];
  string cmd (os.str());
  int fds[2];
  cmd_iofds (fds, cmd);
  ofdstream out (fds[1]);
  ifdinfinistream in (fds[0]);
  in.tie (&out);

  if (opt_init) {
    create_config(in, out, opt_init_dest);
    try {
      nmp.reset(new notmuch_db (opt_notmuch_config, true));
    } catch (whattocatch_t e) { cerr << e.what() << '\n'; exit (1); }
  }
  if (!muchsync_init(nmp->maildir, true))
    exit(1);
  if (!opt_nonew)
    nmp->run_new();
  string dbpath = nmp->maildir + muchsync_dbpath;
  sqlite3 *db = dbopen(dbpath.c_str(), true);
  if (!db)
    exit (1);
  cleanup _c (sqlite3_close_v2, db);

  try {
    muchsync_client (db, *nmp, in, out);
  }
  catch (whattocatch_t &e) {
    cerr << e.what() << '\n';
    exit (1);
  }
}

enum opttag {
  OPT_VERSION = 0x100,
  OPT_SERVER,
  OPT_NOSCAN,
  OPT_UPBG,
  OPT_NOUP,
  OPT_HELP,
  OPT_NONEW,
  OPT_SELF,
  OPT_INIT
};

static const struct option muchsync_options[] = {
  { "version", no_argument, nullptr, OPT_VERSION },
  { "server", no_argument, nullptr, OPT_SERVER },
  { "noscan", no_argument, nullptr, OPT_NOSCAN },
  { "upbg", no_argument, nullptr, OPT_UPBG },
  { "noup", no_argument, nullptr, OPT_NOUP },
  { "noupload", no_argument, nullptr, OPT_NOUP },
  { "nonew", no_argument, nullptr, OPT_NONEW },
  { "init", required_argument, nullptr, OPT_INIT },
  { "self", no_argument, nullptr, OPT_SELF },
  { "config", required_argument, nullptr, 'C' },
  { "help", no_argument, nullptr, OPT_HELP },
  { nullptr, 0, nullptr, 0 }
};

int
main(int argc, char **argv)
{
  umask (077);

  opt_notmuch_config = notmuch_db::default_notmuch_config();
  bool opt_self = false;

  int opt;
  while ((opt = getopt_long(argc, argv, "+C:Fr:s:v",
			    muchsync_options, nullptr)) != -1)
    switch (opt) {
    case 0:
      break;
    case 'C':
      opt_notmuch_config = optarg;
      break;
    case 'F':
      opt_fullscan = true;
      break;
    case 'r':
      opt_remote_muchsync_path = optarg;
      break;
    case 's':
      opt_ssh = optarg;
      break;
    case 'v':
      opt_verbose++;
      break;
    case OPT_VERSION:
      cout << PACKAGE_STRING << '\n';
      exit (0);
    case OPT_SERVER:
      opt_server = true;
      break;
    case OPT_NOSCAN:
      opt_noscan = true;
      break;
    case OPT_UPBG:
      opt_upbg = true;
      break;
    case OPT_NOUP:
      opt_noup = true;
      break;
    case OPT_NONEW:
      opt_nonew = true;
      break;
    case OPT_SELF:
      opt_self = true;
      break;
    case OPT_INIT:
      opt_init = true;
      opt_init_dest = optarg;
      break;
    case OPT_HELP:
      usage(0);
    default:
      usage();
    }

  if (opt_self)
    print_self();
  else if (opt_server) {
    if (opt_init || opt_noup || opt_upbg || optind != argc)
      usage();
    server();
  }
  else if (opt_upbg) {
    int fds[2];
    if (pipe(fds)) {
      cerr << "pipe: " << strerror(errno) << '\n';
      exit (1);
    }
    fcntl(fds[1], F_SETFD, 1);
    if (fork() > 0) {
      char c;
      close(fds[1]);
      read(fds[0], &c, 1);
      if (opt_verbose)
	cerr << "backgrounding\n";
      exit(0);
    }
    close(fds[0]);
    opt_upbg_fd = fds[1];
    client(argc - optind, argv + optind);
  }
  else
    client(argc - optind, argv + optind);
  return 0;
}
